### DISCUSSION TITLE
Woche 5

### DISCUSSION THREAD

**Discussion by TimSchnur** [Christian-2003/CrInGE#18] [Student Exchange]
Hallo zusammen,

in dieser Woche haben wir uns mit der :thinking: **Klassenmodellierung** unseres Projekts beschäftigt.
Dazu haben wir ein sehr komplexes UML-Klassendiagramm erstellt und die Sequenzdiagramme leicht überarbeitet.
Diese Änderungen sollen im laufe der Woche in die :page_with_curl: [**Software Anforderungsspezifikation**](https://github.com/Christian-2003/CrInGE/blob/master/docs/SoftwareAnforderungsspezifikation.md) integriert werden.
Außerdem haben wir die Klassendefinitionen zu den entworfenen Klassen bereits implementiert.

###### Klassendiagramm

![Klassendiagramm](https://github.com/Christian-2003/CrInGE/assets/80711177/5e046608-71f9-47ce-94dd-49139f54de85)

###### geändertes Sequenzdiagramm für den Welteneditor

![Sequenzdiagramm_Welteneditor_Class](https://github.com/Christian-2003/CrInGE/assets/136824971/071a5386-560a-4b94-ab18-5fdca9b4dbfc)

###### Weitere Aufgaben

Neben der Klassenmodellierung haben wir noch einige kleinere Aufgaben durchgeführt:

Unter Anderem haben wir ein SOLID Prinzip :thought_balloon: gewählt, welches wir verstärkt befolgen wollen.
Gewonnen hat das Liskov Substitution Prinzip knapp vor dem Open-Closed Prinzip.

Außerdem haben wir entschieden uns während der Entwicklung aufzuteilen :busts_in_silhouette:.
Die Aufteilung steht noch nicht final fest, allerdings ist der Grundstein die Aufteilung zwischen **Game Editor** und :video_game: **Game Engine**.
Eventuell wird es ein extra Team für das :floppy_disk: Persistieren geben.

Damit wäre auch schon das Ende des Blogs für diese Woche erreicht. Wir freuen uns auch euer Feedback und werden euch nächste Woche weitere Fortschritte berichten.

**Comment by lukpr0** [comment]
Hallo CrInGE Team,
Zu eurem Klassendiagramm stellen sich mir ein paar Fragen:
1) Es gibt eine AssetList. Anhand dessen, dass diese Komponente im Package "view" liegt nehme ich an, dass dies eine Komponente des UIs ist. Falls nicht, was ist deren Zweck? Außerdem wird mir nicht ganz klar, wie die Assets dargestellt werden, gibt es dazu keine Klasse?
2) Warum gibt es keine direkte Assoziation zwischen GameObjectManager und GameObject?
3) Was sind die Aufgaben der Klassen Movement und Velocity?

Bonustipp: Diagramm in höherer Auflösung rendern, man kanns zwar noch lesen aber die Buchstaben werden trotzdem schon von der Auflösung zermatscht

**Reply by TimSchnur** [reply]
Hallo Lukas,
danke für den Tipp. Wir haben bereits geplant das Diagramm Bild in mehrere aufzuteilen, um die Lesbarkeit zu fördern. Bei diesem Zug sollte die Auflösung erhöht werden.

Bezüglich deiner Fragen:
1. Die Klasse AssetList stellt eine Komponente dar, welche die verfügbaren Assets listet. Dabei werden die Assets als Graphik importiert, welche innerhalb des Projektes in der Ordnerstruktur liegt. Somit soll die Klasse ermöglichen alle Assets in der UI verfügbar zu machen.
2. Der Komplette Spielfluss soll momentan über den Game Controller in der Game Loop stattfinden. Dieser wird Methoden des Game Object Managers auf das Game Objekt anwenden. Allerdings benötigen noch mehr Klassen, wie zum Beispiel der Rendermanager, dieses Objekt, weshalb es zentral über die Game Loop verwaltet werden soll.
3. Die Klassen Movement und Velocity werden Berechnungen im Bezug auf Bewegungen und Beschleunigungen übernehmen, sodass man diese leicht über die API verwenden kann. Beispielsweise kann man so als Spieleentwickler die Berechnung der Geschwindigkeit eines fallenden Balles nach 10 Sekunden einfach aus der API nehmen statt selbst ein Script schreiben zu müssen.

Ich hoffe ich konnte dir deine Fragen beantworten und vielen Dank für dein Feedback.
